provider/sgx.py

```diff
@@ -0,0 +1,189 @@
+            epc_size += int(tmp_epc[0].split('-')[1], 16) - int(
+                tmp_epc[0].split("-")[0], 16) + 1
+    return epc_size
+
+
+class SGXTestCapability(object):
+    """
+    Hypervisor sgx test capabilities check.
+    """
+
+    def __init__(self, test, params):
+        """
+
+        :param test: Context of test.
+        :param params: params of running ENV.
+        """
+        self._test = test
+        self._params = params
+
+    def valid_host_sgx_cap(self):
+        """
+        Valid if host sgx enablement
+
+        :return: host epc size/False
+        """
+        try:
+            host_sgx_msg = process.system_output(
+                self._params["host_sgx_check"], shell=True)
+        except Exception as e:
+            self._test.cancel("Host sgx capability check fail %s" % e)
+        else:
+            host_msg_epc_size = _get_epc_size(host_sgx_msg)
+            return host_msg_epc_size
+
+    def valid_host_numa_node(self):
+        """
+        Valid if host numa nodes satisfy test requirement
+
+        :return: True/False
+        """
+        LOG_JOB.debug("Check host's numa node(s)!")
+        host_numa = utils_memory.numa_nodes()
+        node_list = []
+        numa_info = process.getoutput("numactl -H")
+        for i in host_numa:
+            node_size = re.findall(r"node %d size: "
+                                   r"\d+ \w" % i, numa_info)[0].split()[-2]
+            if node_size != '0':
+                node_list.append(str(i))
+        if len(node_list) < int(self._params["monitor_expect_nodes"]):
+            self._test.cancel("host numa nodes %s isn't enough for "
+                              "testing." % node_list)
+
+
+class SGXChecker(object):
+    """
+    General operations for checking guest sgx enablement.
+    """
+
+    def __init__(self, test, params, vm):
+        """
+
+        :param test: Context of test.
+        :param params: params of running ENV.
+        :param vm:VM object.
+        """
+        self._test = test
+        self._params = params
+        self._vm = vm
+        self._monitor = vm.monitor
+
+    def sgx_flags_verify(self, qmp_command, flags):
+        """
+        Check if sgx cpu flags enabled in qmp cmd
+
+        :param qmp_command: query sgx qmp command output
+        :param flags: sgx flags need to be verified
+        """
+        for ele in flags:
+            if qmp_command[ele] is not True:
+                self._test.fail("%s is not enabled, qmp check failed." % ele)
+
+    def host_qmp_sgx_verify(self):
+        """
+        Verify host sgx flags and capability
+        """
+        sgx_flags = self._params['sgx_flags'].split()
+        host_sgx_info = self._monitor.query_sgx_capabilities()
+        self.sgx_flags_verify(host_sgx_info, sgx_flags)
+
+        host_msg_epc_size = self._params["host_epc_size"]
+        host_sgx_section_size = host_sgx_info['section-size']
+        if host_msg_epc_size != int(host_sgx_section_size):
```

Comment:
The size can be different?
e.g. we set 256M epc in bios, but we only configure one 64M epc dev in cfg, then section-size here should be 64M, right?

btw: what's the diffs between query-sgx and query-sgx-capabilities?

Reply:
query_sgx_capabilities used to query host sgx capabilities. query-sgx used to query current VM sgx info. Host section size is configured in BIOS, so if we set 256M epc in bios, then section-size = 256M